Objects/unicodeobject.c

```diff
@@ -9698,11 +9698,11 @@ split(PyObject *self,
     PyObject* out;
     len1 = PyUnicode_GET_LENGTH(self);
     kind1 = PyUnicode_KIND(self);
-    if (maxcount < 0) {
-        maxcount = len1;
-    }
 
-    if (substring == NULL)
+    if (substring == NULL) {
+        if (maxcount < 0) {
+            maxcount = (len1 + 1) / 2;
+        }
```

Comment:
We need to avoid overflow here.

```c
#include <unistd.h>
#include <sys/types.h>
#include <limits.h>
#include <stdio.h>

int main() {
    ssize_t a = SSIZE_MAX;
    printf("%zd\n", (a+1)/2); // -4611686018427387904
    printf("%zd\n", (ssize_t)(((size_t)a+1)/2)); // 4611686018427387904
    return 0;
}
```